File: Coding/Codeforces/C++/actual_first.cpp
#include <iostream>
using namespace std;
int main()
{
	int tests;
	cin>>tests;
	int n;
	for(int i = 0; i < tests; i++)
	{
		int Janswer;
		int afterJanswer;
		cin>>n;
		if(n % 2 == 0)
		{
			for(int j = 1; j < n; j+=2)
			{
				cout<<"? "<<j<<" "<<(j+1)<<endl;
				cin>>Janswer;
				int x;
				cout<<"? "<<(j+1)<<" "<<j<<endl;
				cin>>x;
				Janswer = Janswer*10 + x;
				if(Janswer != 0 && Janswer != 11)
				{
					if(j != 1)
					{
						cout<<"? "<<(j-1)<<" "<<j<<endl;
						cin>>afterJanswer;
						int y;
						cout<<"? "<<j<<" "<<(j-1)<<endl;
						cin>>y;
						afterJanswer = afterJanswer*10 + y;
						if(afterJanswer != 0 && afterJanswer != 11)
							cout<<"! "<<j<<endl;
						else
							cout<<"! "<<(j+1)<<endl;
						break;
					}
					else
					{
						cout<<"? 2 3"<<endl;
						cin>>afterJanswer;
						int y;
						cout<<"? 3 2"<<endl;
						cin>>y;
						afterJanswer = afterJanswer*10 + y;
						if(afterJanswer != 0 && afterJanswer != 11)
							cout<<"! 2"<<endl;
						else
							cout<<"! 1"<<endl;
						break;
					}
				}
			}
		}
		else
		{
			bool done = false;
			for(int j = 1; j < n-1; j+=2)
			{
				cout<<"? "<<j<<" "<<(j+1)<<endl;
				cin>>Janswer;
				int x;
				cout<<"? "<<(j+1)<<" "<<j<<endl;
				cin>>x;
				Janswer = Janswer*10 + x;
				if(Janswer != 0 && Janswer != 11)
				{
					if(j != 1)
					{
						cout<<"? "<<(j-1)<<" "<<j<<endl;
						cin>>afterJanswer;
						int y;
						cout<<"? "<<j<<" "<<(j-1)<<endl;
						cin>>y;
						afterJanswer = afterJanswer*10 + y;
						if(afterJanswer != 0 && afterJanswer != 11)
							cout<<"! "<<j<<endl;
						else
							cout<<"! "<<(j+1)<<endl;
						done = true;
						break;
					}
					else
					{
						cout<<"? 2 3"<<endl;
						cin>>afterJanswer;
						int y;
						cout<<"? 3 2"<<endl;
						cin>>y;
						afterJanswer = afterJanswer*10 + y;
						if(afterJanswer != 0 && afterJanswer != 11)
							cout<<"! 2"<<endl;
						else
							cout<<"! 1"<<endl;
						done = true;
						break;
					}
				}
			}
			if(!done)
				cout<<"! "<<n<<endl;
		}
	}
}

File: Coding/Codeforces/Python/actual_first.py
tests = int(input())
n = 0
for i in range(tests):
	Janswer = 0
	afterJanswer = 0
	n = int(input())
	if n % 2 == 0:
		j = 1
		while j < n:
			Janswer = int(input(f"? {j} {j+1}\n"))
			x = int(input(f"? {j+1} {j}\n"))
			Janswer = Janswer*10 + x
			if Janswer != 0 and Janswer != 11:
				if j != 1:
					afterJanswer = int(input(f"? {j-1} {j}\n"))
					y = int(input(f"? {j} {j-1}\n"))
					afterJanswer = afterJanswer*10 + y
					if afterJanswer != 0 and afterJanswer != 11:
						print(f"! {j}")
					else:
						print(f"! {j+1}")
					break
				else:
					afterJanswer = int(input("? 2 3\n"))
					y = int(input("? 3 2\n"))
					afterJanswer = afterJanswer*10 + y
					if afterJanswer != 0 and afterJanswer != 11:
						print("! 2")
					else:
						print("! 1")
					break
			j += 2
	else:
		done = False
		j = 1
		while j < n - 1:
			Janswer = int(input(f"? {j} {j+1}\n"))
			x = int(input(f"? {j+1} {j}\n"))
			Janswer = Janswer*10 + x
			if Janswer != 0 and Janswer != 11:
				if j != 1:
					afterJanswer = int(input(f"? {j-1} {j}\n"))
					y = int(input(f"? {j} {j-1}\n"))
					afterJanswer = afterJanswer*10 + y
					if afterJanswer != 0 and afterJanswer != 11:
						print(f"! {j}")
					else:
						print(f"! {j+1}")
					done = True
					break
				else:
					afterJanswer = int(input("? 2 3\n"))
					y = int(input("? 3 2\n"))
					afterJanswer = afterJanswer*10 + y
					if afterJanswer != 0 and afterJanswer != 11:
						print("! 2")
					else:
						print("! 1")
					done = True
					break
			j += 2
		if not done:
			print(f"! {n}")